Source/Dafny/Compilers/Compiler-Csharp.cs

```diff
@@ -1923,6 +2011,16 @@ protected class ClassWriter : IClassWriter {
         return IdProtect(udt.CompileName);
       }
 
+      //Use the interface if applicable (not handwritten, or incompatible variance)
+      bool compileIt = true;
+      if ((cl is IndDatatypeDecl || cl is CoDatatypeDecl)
```

Comment:
You can replace `(cl is IndDatatypeDecl || cl is CoDatatypeDecl)` with just `cl is DatatypeDecl`.